pkg/proto/addresses.go

```diff
@@ -4,192 +4,323 @@ import (
 	"bytes"
 	"encoding/binary"
 	"fmt"
+	"golang.org/x/crypto/sha3"
 	"io"
 	"strconv"
 	"strings"
 
 	"github.com/mr-tron/base58/base58"
 	"github.com/pkg/errors"
+	"github.com/umbracle/fastrlp"
 	"github.com/wavesplatform/gowaves/pkg/crypto"
 	"github.com/wavesplatform/gowaves/pkg/errs"
 	g "github.com/wavesplatform/gowaves/pkg/grpc/generated/waves"
 	"github.com/wavesplatform/gowaves/pkg/libs/serializer"
 )
 
 const (
-	headerSize     = 2
-	bodySize       = 20
-	checksumSize   = 4
-	AddressSize    = headerSize + bodySize + checksumSize
-	aliasFixedSize = 4
+	AddressIDSize            = 20
+	wavesAddressHeaderSize   = 2
+	wavesAddressBodySize     = AddressIDSize
+	wavesAddressChecksumSize = 4
+	WavesAddressSize         = wavesAddressHeaderSize + wavesAddressBodySize + wavesAddressChecksumSize
 
-	addressVersion byte = 0x01
-	aliasVersion   byte = 0x02
+	wavesAddressVersion byte = 0x01
+	aliasVersion        byte = 0x02
 
+	aliasFixedSize = 4
 	AliasMinLength = 4
 	AliasMaxLength = 30
 	AliasAlphabet  = "-.0123456789@_abcdefghijklmnopqrstuvwxyz"
 	AliasPrefix    = "alias"
 
+	EthereumAddressSize = AddressIDSize
+
 	MainNetScheme   byte = 'W'
 	TestNetScheme   byte = 'T'
 	StageNetScheme  byte = 'S'
 	CustomNetScheme byte = 'E'
 )
 
-// Address is the transformed Public Key with additional bytes of the version, a blockchain scheme and a checksum.
-type Address [AddressSize]byte
+type AddressID [AddressIDSize]byte
+
+func (a AddressID) Bytes() []byte {
+	return a[:]
+}
+
+type Address interface {
+	ID() AddressID
+	Bytes() []byte
+	String() string
+}
+
+// EthereumAddress is the first 20 bytes of Public Key's hash for the Waves address, or the 20 bytes of an Ethereum address.
+type EthereumAddress [EthereumAddressSize]byte
+
+// BytesToEthereumAddress returns EthereumAddress with value b.
+// If b is larger than len(h), b will be cropped from the left.
+func BytesToEthereumAddress(b []byte) EthereumAddress {
```

Comment:
It needs to have a simple check on length of `b` like `if len(b) == 20`

Reply:
I think this is unnecessarily. The main point is
> If b is larger than len(h), b will be cropped from the left.

Reply:
It reserved for future. See #529